app/src/main/java/com/yello/presentation/auth/SignInActivity.kt

```diff
@@ -0,0 +1,122 @@
+package com.yello.presentation.auth
+
+import android.content.Intent
+import android.os.Bundle
+import com.example.ui.base.BindingActivity
+import com.example.ui.view.setOnSingleClickListener
+import com.kakao.sdk.auth.model.OAuthToken
+import com.kakao.sdk.common.model.ClientError
+import com.kakao.sdk.common.model.ClientErrorCause
+import com.kakao.sdk.common.util.Utility
+import com.kakao.sdk.user.UserApiClient
+import com.yello.R
+import com.yello.databinding.ActivitySignInBinding
+import timber.log.Timber
+
+class SignInActivity : BindingActivity<ActivitySignInBinding>(R.layout.activity_sign_in) {
+
+    private lateinit var appLoginCallback: (OAuthToken?, Throwable?) -> Unit
+    private lateinit var accountLoginCallback: (OAuthToken?, Throwable?) -> Unit
+    private lateinit var serviceTermsList: List<String>
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        // TODO : 4명 키 해시 모두 받은 다음에 코드 지우기
+        val keyHash = Utility.getKeyHash(this)
+        Timber.tag("signIn").d(keyHash)
+
+        binding.btnSignIn.setOnSingleClickListener {
+            setServiceTerms()
+            setAppLoginCallback()
+            setAccountLoginCallback()
+            startKakaoLogin()
+            getUserInfo()
+        }
+    }
+
+
+    // TODO : 카카오 로그인 동의 화면에 포함할 서비스 약관 항목 지정 (기획 측에서 필요 약관 확정해서 넘겨주기)
+    private fun setServiceTerms() {
+        serviceTermsList = listOf("profile_nickname", "profile_image", "account_email", "age_range", "friends")
+    }
+
+    // 웹에서 계정 로그인 callback 구성
+    private fun setAccountLoginCallback() {
+        accountLoginCallback = { token, error ->
+            if (error != null) {
+                Timber.tag("signIn").e(error, "카카오계정으로 로그인 실패")
+            } else if (token != null) {
+                postKakaoAccessToken(token)
+                startSocialSyncActivity()
+            }
+        }
+    }
+
+    private fun loginWithAccountCallback() {
+        UserApiClient.instance.loginWithKakaoAccount(
+            context = binding.root.context,
+            callback = appLoginCallback,
+            serviceTerms = serviceTermsList
+        )
+    }
+
+    // 카카오톡 앱 로그인 callback 구성
+    private fun setAppLoginCallback() {
+        appLoginCallback = { token, error ->
+            if (error != null) {
+                Timber.tag("signIn").e(error, "카카오톡으로 로그인 실패")
+
+                // 뒤로가기 경우 예외 처리
+                if (error is ClientError && error.reason == ClientErrorCause.Cancelled) {
+                    Timber.tag("signIn").e(error, "유저가 로그인 취소")
+                } else {
+                    // 카카오톡 연결 실패 시, 계정으로 로그인 시도
+                    loginWithAccountCallback()
+                }
+
+            } else if (token != null) {
+                postKakaoAccessToken(token)
+                startSocialSyncActivity()
+            }
```

Comment:
```suggestion
            error ?: return
            Timber.tag("signIn").e(error, "카카오톡으로 로그인 실패")
             if (error is ClientError && error.reason == ClientErrorCause.Cancelled) {
                  Timber.tag("signIn").e(error, "유저가 로그인 취소")
                  return
              }
              loginWithAccountCallback()
              token ?: return
              postKakaoAccessToken(token)
              startSocialSyncActivity()
```
코드 스타일마다 다를 수 있는데 early return 패턴을 활용하면 if문이 중첩되지 않도록 작성할 수도 있습니당

Reply:
저도 안쪽 if문에서 return을 활용해보고자 했었는데 계속 빨간줄 뜨길래 일단 저렇게 해뒀어요 !
지금 Suggested change를 붙여넣어봐도 'return' is not allowed in here 로 뜨네요 ㅜㅜ

Reply:
아 람다식은 익명함수이기 때문에 label을 직접 지정해줘야 early return이 가능하네요!

```
accountLoginCallback = loginCallback@{ token, error ->
    error ?: return@loginCallback
    Timber.tag("signIn").e(error, "카카오톡으로 로그인 실패")
    if (error is ClientError && error.reason == ClientErrorCause.Cancelled) {
        Timber.tag("signIn").e(error, "유저가 로그인 취소")
        return@loginCallback
    }
    loginWithAccountCallback()
    token ?: return@loginCallback
    postKakaoAccessToken(token)
    startSocialSyncActivity()
}
```

early return 패턴을 적용하고 싶으시다면 이렇게 할 수 있을 것 같습니당